### PR TITLE
Fix crash in spans when time() is not monotonic

### PR DESCRIPTION
**Pull request by crusaderky**
Fix crash:
```python
  File "/opt/coiled/env/lib/python3.11/site-packages/distributed/spans.py", line 328, in cumulative_worker_metrics
    unknown_seconds = max(0.0, self.active_cpu_seconds - known_seconds)
                               ^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/distributed/spans.py", line 439, in active_cpu_seconds
    return sum((t1 - t0) * nthreads for t0, t1, nthreads in self.nthreads_intervals)
                                                            ^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/distributed/spans.py", line 410, in nthreads_intervals
    is_active_t, is_active_flag = zip(*self._active_timeseries())
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/coiled/env/lib/python3.11/site-packages/distributed/spans.py", line 382, in _active_timeseries
    assert delta > 0
           ^^^^^^^^^
AssertionError
```

This was triggered when two subsequent calls to `time()` are not strictly monotonic ascending.
No unit tests, as I'm not too eager to monkey-patch `time()` to reproduce

### CODE REVIEW

**Comment by ntabris** [comment]
Mostly just curious but if the problem is that `time()` isn't monotonic, why isn't the solution to use a monotonic clock?

**Comment by crusaderky** [comment]
> Mostly just curious but if the problem is that `time()` isn't monotonic, why isn't the solution to use a monotonic clock?

Because `time.monotonic()` is not a wall clock, but just the number of seconds since process start.
In Windows we tamper with `time.time()` due to low resolution. In theory, we could do the same in Linux to create a monotonic wall time. In practice, it would be ugly as hell (see `distributed.metrics._WindowsTime`), slow, and very bug-prone.
So I feel it's much better to just work within the constraints of the standard UNIX syscalls.

File: distributed/spans.py [hunk 1]
```diff
@@ -364,26 +366,25 @@ def _active_timeseries(self) -> Iterator[tuple[float, bool]]:
...
-            events.append((child.stop if child.done else now, -1))
+            events += [(child.enqueued, 1), (child.stop, -1)]
         events.sort()
 
         n_active = 0
         for t, delta in events:
-            if not n_active:
-                assert delta > 0
+            # Note: in case of identical timestamps, there may be loops after sorting
+            # were n_active < 0
```
Comment (hendrikmakait) [inline distributed/spans.py:382]: IIUC, `child.enqueued <= child.stop`. How about sorting by timestamp only `events.sort(key=lambda event: event[0])`? This should maintain relative ordering and avoid `n_active < 0`. (It took me a moment to understand why `n_active` would ever drop below 0 here.)

Comment (crusaderky) [reply distributed/spans.py]: Done

**Pull request merged**